### PR TITLE
Support `spark.sql.parquet.datetimeRebaseModeInWrite=LEGACY`

### PR DESCRIPTION
**Pull request by ttnghia**
Closes:
 * https://github.com/NVIDIA/spark-rapids/issues/8596.

Depends on:
 * https://github.com/NVIDIA/spark-rapids-jni/pull/1375
 * https://github.com/rapidsai/cudf/pull/14026

TBA.

### CODE REVIEW

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala [hunk 1]
```diff
@@ -131,7 +133,13 @@ object GpuParquetFileFormat {
     }
 
     SparkShimImpl.parquetRebaseWrite(sqlConf) match {
-      case "EXCEPTION" | "CORRECTED"| "LEGACY" => // Good
+      case "EXCEPTION" | "CORRECTED" => // Good
+      case "LEGACY" =>
+        val tz = TimeZone.getDefault.getID
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala:138]: nit: can we make this code match the other timezone check code?

https://github.com/NVIDIA/spark-rapids/blob/feb0c86d4cdbed747335a1db4551422bbad926ca/sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala#L830-L842

I am not saying that you need to call these methods, but I am saying that we should use the same code to check the timestamp is UTC.

File: integration_tests/src/main/python/parquet_write_test.py [hunk 2]
```diff
@@ -70,6 +70,13 @@
                                                     MapGen(StringGen(pattern='key_[0-9]', nullable=False),
                                                            simple_string_to_string_map_gen)]
 
+parquet_datetime_gen_simple = [DateGen(), TimestampGen()]
```
Comment (revans2) [inline integration_tests/src/main/python/parquet_write_test.py:73]: Python itself has some issues with dates/timestamps in the far future. Do we need to adjust these so we don't run into issues?  Do we want some tests that explicitly test rebase where we know that the range of dates/timestamps all need to be rebased (or at least it is really likely that they need to be rebased).

Comment (ttnghia) [reply integration_tests/src/main/python/parquet_write_test.py:73]: > Python itself has some issues with dates/timestamps in the far future.

I didn't know that. Of course we can limit the `end` date/timestamp as there will be not any rebase after `1582-10-15` so the test range from `1-1-1` to some day after `1582-10-15` should be good.

Comment (revans2) [reply integration_tests/src/main/python/parquet_write_test.py:73]: Some version of python cannot support > 9999 for the year

https://docs.python.org/3/library/datetime.html#datetime.MAXYEAR

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

**Pull request merged**